### PR TITLE
Rework implicit allocation in constructors

### PR DESCRIPTION
**Pull request by dcodeIO**
As reported in https://github.com/AssemblyScript/assemblyscript/issues/1225 some weirdness is happening in constructors, caused by late `this` allocations. While investigating, it turned out that this mechanism is connected to quite a few other assumptions made by inlining and reference counting, so I decided to implement a new but more conservative mechanism.

With the change, the allocation is injected right at the start of the constructor as soon as any code path requires it (falls-through or accesses `this`). As a side-effect, the following special cases are no longer allowed:

* A constructor with an explicit return must not access `this`. Not yet sure if this can become a warning instead of an error.
* A class with an explicitly returning constructor must be sealed since `super` calls in this scenario will break due to the resulting `this` mismatch.

On the upside, implicit allocations in constructors are now guaranteed to be performed exactly once, reducing code size in scenarios where `this` is first accessed in branches.

fixes https://github.com/AssemblyScript/assemblyscript/issues/1225

### CODE REVIEW

File: std/assembly/array.ts [hunk 1]
```diff
@@ -111,7 +111,7 @@ export class Array<T> {
     return value;
   }
 
-  @unsafe @operator("{}") private __unchecked_get(index: i32): T {
+  @unsafe @operator("{}") private __uget(index: i32): T {
```
Comment (MaxGraey) [inline std/assembly/array.ts:114]: that's why latest ts-node turn off cache completely

Comment (dcodeIO) [reply std/assembly/array.ts:114]: I've encountered the cache problem earlier with `__unchecked_get`, and just then decided to unify the name here. From what I can tell, the ts-node caching stuff typically goes haywire as soon as `CommonFlags`, `FlowFlags` or `LocalFlags` are changed with new values.

**Comment by dcodeIO** [comment]
This now generally allows returning in constructors as long as `this` is returned. If something else than `this` is returned somewhere, the class must be `@sealed`. In any case, if something else than `this` is returned but `this` is being accessed somewhere in the constructor, a pedantic diagnostic is emitted hinting at the potentially dropped `this` allocation.

**Comment by dcodeIO** [comment]
Also renamed `@sealed` to `@final` now, so we can reuse it for methods or anything else final as well eventually.

File: std/assembly/array.ts (hunk 1 above)
Comment (dcodeIO) [reply std/assembly/array.ts:114]: In this particular case, the problem seems to occur when the value of `FlowFlags.UNCHECKED_CONTEXT` changes. This one just so happens to be the last flag in that enum, so is pretty much always affected. Good indicator I guess :)

**Pull request merged**